File: documentation-examples-kotlin/src/kotlin/dev/typr/foundationskt/docs/core/PersistedCodecs.kt
package dev.typr.foundationskt.docs.core

import dev.typr.foundationskt.*
import dev.typr.foundationskt.data.*
import dev.typr.foundationskt.docs.core.PersistedTypes.*

@Suppress("unused")
class PersistedCodecs {
    val venueIdType: DuckDbType<VenueId> = DuckDbTypes.bigint.transform(::VenueId, VenueId::value)

    //start
    val venueCodec: RowCodecNamed<Venue> =
        RowCodec.namedBuilder<Venue>()
            .field("name", DuckDbTypes.varchar, Venue::name)
            .field("capacity", DuckDbTypes.integer, Venue::capacity)
            .build(::Venue)

    val persistedVenueCodec: RowCodecNamed<PersistedVenue> =
        RowCodec.ofNamed("id", venueIdType)
            .join(venueCodec)
            .to({ (id, venue) -> PersistedVenue(id, venue) }, { Pair(it.id, it.venue) })
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/PersistedCodecs.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.data.*
import dev.typr.foundationssc.docs.core.PersistedTypes.*

@SuppressWarnings(Array("unused"))
object PersistedCodecs:
  val venueIdType: DuckDbType[VenueId] = DuckDbTypes.bigint.transform(VenueId.apply, _.value)

  // start
  val venueCodec: RowCodecNamed[Venue] = RowCodec
    .namedBuilder[Venue]()
    .field("name", DuckDbTypes.varchar)(_.name)
    .field("capacity", DuckDbTypes.integer)(_.capacity)
    .build(Venue.apply)

  val persistedVenueCodec: RowCodecNamed[PersistedVenue] =
    RowCodec
      .ofNamed("id", venueIdType)
      .join(venueCodec)
      .to(PersistedVenue.apply, pv => (pv.id, pv.venue))
  // stop
